### PR TITLE
r.buildvrt.gdal: Create GDAL VRTs for GDAL linked raster maps

### PR DESCRIPTION
**Pull request by ninsbl**
A workaround for https://github.com/OSGeo/grass/issues/4345

### CODE REVIEW

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.html [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+maps and links them to the mapset with <em>r.external</em>. The module is
+written as a workaround for a limitation in GRASS GIS Virtual Rasters (VRT)
+format with GDAL-linked raster maps (through <em>r.external</em> /
+<em>r.external.out</em>. In that case GRASS GIS Virtual Rasters currently
+show performance issues. See:
+<a href="https://github.com/OSGeo/grass/issues/4345">#4345</a>
+
+<p>
+For the resulting maps GDAL VRT text files are created either in a
+directory named "gdal" in the current mapset og in a userdefined <b>
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.html]:
```suggestion
directory named "gdal" in the current mapset or in a user-defined <b>
```

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                return str(gdal_path)
+
+        # Check if Path can be read from command history
+        gdal_path = Path(
+            gs.parse_key_val(
+                gs.parse_command("r.info", flags="e", map=map_name)["comments"]
+                .replace("\\", "")
+                .replace('"', ""),
+                vsep=" ",
+            )["input"]
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Please look at https://github.com/OSGeo/grass/pull/4216 whether that would help here.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            / "gdal"
+        )
+        if header_path.exists():
+            gdal_path = Path(
+                gs.parse_key_val(header_path.read_text().replace(": ", "="))["file"]
+            )
+            if gdal_path.exists():
+                return str(gdal_path)
+
+        # Check if Path can be read from command history
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Don't use capital letter for Path and instead say which path that is. This is probably the most fragile part of the code in terms of relying on other things to be in place.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            Path(gis_env["GISDBASE"])
+            / gis_env["LOCATION_NAME"]
+            / map_info["mapset"]
+            / "cell_misc"
+            / map_info["name"]
+            / "gdal"
+        )
+        if header_path.exists():
+            gdal_path = Path(
+                gs.parse_key_val(header_path.read_text().replace(": ", "="))["file"]
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Would the following work?

```suggestion
                gs.parse_key_val(header_path.read_text(), sep=": "))["file"]
```

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                vsep=" ",
+            )["input"]
+        )
+        if gdal_path.exists():
+            return str(gdal_path)
+
+    # Get native GRASS GIS format header
+    gdal_path = Path(gs.find_file(map_name)["file"].replace("/cell/", "/cellhd/"))
+    if gdal_path.exists():
+        return gdal_path
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: I'm not sure how to handle that, but I did hit this issue in the past. It seems we should have API to make usage of GRASS Driver easier. Perhaps spit a GDAL-compatible path out in r.info output?

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    try:
+        from osgeo import gdal
+    except ImportError:
+        gs.fatal(
+            _(
+                "Unable to load GDAL Python bindings (requires "
+                "package 'python-gdal' or Python library GDAL "
+                "to be installed)."
+            )
+        )
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Given that you are using gdal only in the _main_ function, moving this to _main_ should be easy and you will avoid warnings about options and flags being on the top-level.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    # Create GDAL VRT
+    vrt_path = str(vrt_dir / f"{output}.vrt")
+    gs.verbose(_("Creating GDAL VRT '{}'.").format(vrt_path))
+    gdal.BuildVRT(vrt_path, inputs)
+
+    # Import (link) GDAL VRT
+    gs.run_command(
+        "r.external",
+        quiet=True,
+        overwrite=gs.overwrite(),
```
Comment (wenzeslaus) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: This should be automatic.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
This looks great. Thanks for sharing it. I have some minor comments.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.html [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+time r.univar map=vrt_${format_type}_gdal
+
+</pre></div>
+
+<h2>SEE ALSO</h2>
+<em>
+<a href="https://grass.osgeo.org/grass-stable/manuals/r.buildvrt.html">r.buildvrt</a>,
+<a href="https://grass.osgeo.org/grass-stable/manuals/r.patch.html">r.patch</a>,
+<a href="https://grass.osgeo.org/grass-stable/manuals/r.external.html">r.external</a>,
+<a href="https://grass.osgeo.org/grass-stable/manuals/r.external.out.html">r.external.out</a>,
```
Comment (neteler) [inline src/raster/r.buildvrt.gdal/r.buildvrt.gdal.html]:
```suggestion
time r.univar map=vrt_${format_type}_gdal
</pre></div>

<h2>SEE ALSO</h2>
<em>
<a href="https://grass.osgeo.org/grass-stable/manuals/r.buildvrt.html">r.buildvrt</a>,
<a href="https://grass.osgeo.org/grass-stable/manuals/r.patch.html">r.patch</a>,
<a href="https://grass.osgeo.org/grass-stable/manuals/r.external.html">r.external</a>,
<a href="https://grass.osgeo.org/grass-stable/manuals/r.external.out.html">r.external.out</a>
```

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 5 above)
Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Sounds good.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 6 above)
Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Implemented in [bd4e206](https://github.com/OSGeo/grass-addons/pull/1209/commits/bd4e2061a73f6acf4d5f314abdb7cd7112da8088)

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 2 above)
Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: JSON output now used in [bd4e206](https://github.com/OSGeo/grass-addons/pull/1209/commits/bd4e2061a73f6acf4d5f314abdb7cd7112da8088)

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 3 above)
Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Indeed. This is what I started with when I tried to get the GDAL-path of a linked dataset, but I am not 100% this is reached in normal cases of GDAL-linked data. So we may remove it...

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 7 above)
Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Addressed in [bd4e206](https://github.com/OSGeo/grass-addons/pull/1209/commits/bd4e2061a73f6acf4d5f314abdb7cd7112da8088)

**Comment by ninsbl** [comment]
The requested changes should be addressed.

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 5 above)
Comment (wenzeslaus) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Would you mind opening an issue for this?

**Review by wenzeslaus — APPROVED** [review]

File: src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py (hunk 3 above)
Comment (wenzeslaus) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: True, I think if the gdal header information is missing, then there is really a big problem and the data can't be actually read, so you can remove it without loosing any functionality.

Comment (ninsbl) [reply src/raster/r.buildvrt.gdal/r.buildvrt.gdal.py]: Removed in [ea2a29d](https://github.com/OSGeo/grass-addons/pull/1209/commits/ea2a29ddcde6073d00dfa0817852e44733ab8a3e)

**Pull request merged**